### PR TITLE
Documentation for external client_id and client_secret.

### PR DESCRIPTION
**Pull request by danielzeller-okta**
Primary Reviewers: @larsjohansen-okta @richardmateosian-okta 

@federations-okta @mystiberry-okta 

OKTA-113371

### CODE REVIEW

File: _source/_docs/api/resources/oauth-clients.md [hunk 1]
```diff
@@ -419,7 +421,7 @@ Parameter | Description                        | ParamType | DataType
 clientId  | `clientId` of a specific client    | URL       | String                                 | TRUE     |
 settings  | OAuth client registration settings | Body      | [Client Settings](#oauth-client-model) | TRUE     |
 
-> All settings must be specified when updating a client application, **partial updates are not supported!**
+> All settings must be specified when updating a client application, **partial updates are not supported!** The one exception to this is the `client_secret` which will not be updated if omitted from an update.
```
Comment (richardmateosian-okta) [inline _source/_docs/api/resources/oauth-clients.md]: This needs a comment before "which." I'd consider rewriting the whole thing as: 

> Partial updates are not supported, except that the client secret is immutable and should not be specified in updates. 

You might also want to specify what happens if you submit an update with missing settings.

File: _source/_docs/api/resources/oauth-clients.md [hunk 2]
```diff
@@ -72,12 +72,14 @@ Client applications have the following properties:
 | _links                     | discoverable resources related to the app                         | [JSON HAL](http://tools.ietf.org/html/draft-kelly-json-hal-06)         | TRUE     | FALSE  | TRUE     | 
 |----------------------------+-------------------------------------------------------------------+------------------------------------------------------------------------+----------+--------+----------|
 
-> The `client_id`, `client_id_issued_at`, and `client_secret` attributes are only available after a client is created.
+> The `client_id` may be specified during the initial creation of a client application after which the value is immutable. If no `client_id` is provided Okta will generate one for the client application.
```
Comment (richardmateosian-okta) [inline _source/_docs/api/resources/oauth-clients.md]: I'd rewrite this as:

When creating a client application, you can specify `client_id` or let Okta generate a value for it. Thereafter, `client_id` is immutable.

File: _source/_docs/api/resources/oauth-clients.md [hunk 3]
```diff
@@ -72,12 +72,14 @@ Client applications have the following properties:
...
 
-> Client secret is only shown on the initial creation of a client application (and only if the token_endpoint_auth_method is one that requires a client secret).
-  It is never returned in a GET call.
+> The `client_secret` is only shown on the initial creation of a client application (and only if the `token_endpoint_auth_method` is one that requires a client secret).
+  It is never returned in a GET call. If a `client_secret` is not provided on creation and the `token_endpoint_auth_method` requires one Okta will generate a random `client_secret` for the client application.
 
-> The `service` application type and `client_credentials` grant type [a {% api_lifecycle beta%} feature](/docs/api/getting_started/releases-at-okta.html).
+> The `client_id` and `client_secret` may only be comprised of printable characters. The precise set of permissible characters is defined in [Appendix A of the OAuth 2.0 Spec](https://tools.ietf.org/html/rfc6749#appendix-A). Both values may contain at most 100 characters, and the `client_secret` must contain at least 14 characters.
```
Comment (richardmateosian-okta) [inline _source/_docs/api/resources/oauth-clients.md]: Change "may only be comprised of" to "must consist of."

File: _source/_docs/api/resources/oauth-clients.md [hunk 4]
```diff
@@ -72,12 +72,14 @@ Client applications have the following properties:
...
+
+> The `service` application type and `client_credentials` grant type [are {% api_lifecycle beta%} features](/docs/api/getting_started/releases-at-okta.html).
 
 > The "grant_types" and "response_types" values described above are partially orthogonal, as they refer to arguments passed to different
```
Comment (richardmateosian-okta) [inline _source/_docs/api/resources/oauth-clients.md]: I'm not sure what you mean to convey by quotes in this paragraph, but I think they should be back ticks.

File: _source/_docs/api/resources/oauth-clients.md [hunk 5]
```diff
@@ -72,12 +72,14 @@ Client applications have the following properties:
...
-  It is never returned in a GET call.
+> The `client_secret` is only shown on the initial creation of a client application (and only if the `token_endpoint_auth_method` is one that requires a client secret).
+  It is never returned in a GET call. If a `client_secret` is not provided on creation and the `token_endpoint_auth_method` requires one Okta will generate a random `client_secret` for the client application.
 
-> The `service` application type and `client_credentials` grant type [a {% api_lifecycle beta%} feature](/docs/api/getting_started/releases-at-okta.html).
+> The `client_id` and `client_secret` may only be comprised of printable characters. The precise set of permissible characters is defined in [Appendix A of the OAuth 2.0 Spec](https://tools.ietf.org/html/rfc6749#appendix-A). Both values may contain at most 100 characters, and the `client_secret` must contain at least 14 characters.
+
+> The `service` application type and `client_credentials` grant type [are {% api_lifecycle beta%} features](/docs/api/getting_started/releases-at-okta.html).
```
Comment (yuliu-okta) [inline _source/_docs/api/resources/oauth-clients.md]: this can be removed, API Access Management now is EA.

Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth-clients.md]: already removed in a different PR from Karl.

File: _source/_docs/api/resources/oauth-clients.md (hunk 4 above)
Comment (danielzeller-okta) [inline _source/_docs/api/resources/oauth-clients.md]: These lines being white in the github UI indicates that I did not make any modifications to them. That being said I've changed the quotes to back ticks

File: _source/_docs/api/resources/oauth-clients.md (hunk 1 above)
Comment (danielzeller-okta) [inline _source/_docs/api/resources/oauth-clients.md]: The client secret is not immutable and does not NEED to be specified in updates. I've added the comment about what happens if an update is missing settings.

File: _source/_docs/api/resources/oauth-clients.md [hunk 6]
```diff
@@ -72,17 +72,19 @@ Client applications have the following properties:
 | _links                     | discoverable resources related to the app                         | [JSON HAL](http://tools.ietf.org/html/draft-kelly-json-hal-06)         | TRUE     | FALSE  | TRUE     |
 |----------------------------+-------------------------------------------------------------------+------------------------------------------------------------------------+----------+--------+----------|
 
-> The `client_id`, `client_id_issued_at`, and `client_secret` attributes are only available after a client is created.
+> When creating a client application, you can specify the `client_id` or let Okta generate a vlue for it. Thereafter, the `client_id` is immutable.
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/oauth-clients.md]: vlue -> value

**Comment by larsjohansen-okta** [comment]
It looks ok to me. 🚢

**Comment by richardmateosian-okta** [comment]
Ditto what Lars said. Looks good to me once the typo is fixed

**Comment by richardmateosian-okta** [comment]
Merging the old-fashioned way

**Pull request merged**